### PR TITLE
Feature/1042/cursor indicator for move and rotate

### PR DESCRIPTION
**Pull request by JaiJoPa**
# Cursor changes its Icon according to the mouse action

_**Please read the [CONTRIBUTING.md](https://github.com/MaibornWolff/codecharta/blob/main/CONTRIBUTING.md) before opening a PR.**_

closes: #1042 

## Description

 - The mouse is showing : 
      - the [grabbing](https://www.w3schools.com/cssref/playit.asp?filename=playcss_cursor&preval=grabbing) icon when double-clicking on the map
      - the [move](https://www.w3schools.com/cssref/playit.asp?filename=playcss_cursor&preval=move) icon when right clicking the map to show that it can be moved
     - the  [pointer](https://www.w3schools.com/cssref/playit.asp?filename=playcss_cursor&preval=pointer) icon when hovering over the view cube
- A static function changeCursorIndicator is added which can be given the wanted mouse indicator(grabbing, moving, default...)
- On mouseDown-right, it changes the icon to moving
- On mouseDown-left the second time, it changes the icon to grabbing
- On mouseUp it changes the icon to default
-  The pointer is showed when the  triggerViewCubeHoverEvent function is called and the icon is returning back to default when triggerViewCubeUnhoverEvent is called
- We keep track of the mouse position to prevent the map from opening a link to a building or the sidebar when releasing the left-button on a building after moving the map 

## Screenshots or gifs

MouseIndicator

### CODE REVIEW

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts [hunk 1]
```diff
@@ -36,6 +36,13 @@ export enum ClickType {
 	RightClick = 2
 }
 
+export enum CursorType{
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts:39]: seems like you're not running the pre-commit hooks. Please run `npm i` on the root and activate `Run Git-Hooks` in IntelliJ. Otherwise, run `npm run format` from the root.

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts [hunk 2]
```diff
@@ -308,6 +308,25 @@ describe("codeMapMouseEventService", () => {
 		})
 	})
 
+	describe("changeCursorIndicator", ()=>{
+		it("should set the mouseIcon to grabbing, pointer, moving and default", () =>{
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts:312]: test description is wrong.

```suggestion
it("should set the mouseIcon to grabbing", () =>{
```

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts [hunk 3]
```diff
@@ -316,13 +335,15 @@ describe("codeMapMouseEventService", () => {
...
-			it("should not do anything when no building is hovered and nothing is selected", () => {
+			it("should not do anything when no building is hovered and nothing is selected, but should call changeCursorIndicator with default", () => {
+				CodeMapMouseEventService.changeCursorIndicator = jest.fn()
 				threeSceneService.getHighlightedBuilding = jest.fn()
 				threeSceneService.getSelectedBuilding = jest.fn()
 
 				codeMapMouseEventService.onDocumentMouseUp(event)
 
 				expect(threeSceneService.selectBuilding).not.toHaveBeenCalled()
+				expect(CodeMapMouseEventService.changeCursorIndicator).toHaveBeenCalledWith(CursorType.Default)
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts:346]: Instead of checking if the method is called, do something more intuitive and check if  `document.body.style.cursor` has been set correctly. Mocking methods is useful, when they're from another file / service.

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts [hunk 4]
```diff
@@ -187,7 +202,7 @@ export class CodeMapMouseEventService
 
 	private onLeftClick() {
 		this.threeSceneService.clearSelection()
-		if (this.intersectedBuilding) {
+		if (this.intersectedBuilding && !this.hasMouseMoved(this.mouseOnLastClick)) {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts:205]: Why is this required?

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts (hunk 3 above)
Comment (NearW) [reply visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.spec.ts:346]: This applies to the other tests as well.

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts (hunk 4 above)
Comment (NearW) [reply visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts:205]: > We keep track of the mouse position to prevent the map from opening a link to a building or the sidebar when releasing the left-button on a building after moving the map

I should've read your wonderful PR description. Thanks for that.
Nonetheless: My question here might come up later again, when someone looks at the code. Please add a small comment over the condition, explaining why this is necessary.

**Comment by NearW** [comment]
Remember to add labels and update the `Issue: #` to `fixes: #` or `closes: #`

**Comment by RomanenkoVladimir** [comment]
I checked it out and it seems to work perfectly, however what do you think about using e2e testing? 
That would ensure the correctness of the interaction as well as the correction of called events

File: visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts [hunk 5]
```diff
@@ -201,8 +206,7 @@ export class CodeMapMouseEventService
 
 	private onLeftClick() {
 		this.threeSceneService.clearSelection()
-		// track mouse position to prevent opening a link to a building or the sidebar when releasing on a building after moving the map
-		if (this.intersectedBuilding && !this.hasMouseMoved(this.mouseOnLastClick)) {
+		if (this.intersectedBuilding) {
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/codeMap/codeMap.mouseEvent.service.ts]: That's completely right! I don't remember taking that out, but seems like I did. Good that you've seen that !

**Review by NearW — APPROVED** [review]
Move [this comment](https://github.com/MaibornWolff/codecharta/pull/1164#pullrequestreview-472375204) to it's own issue.
Mergable afterwards.

**Comment by JaiJoPa** [comment]
> Move [this comment](https://github.com/MaibornWolff/codecharta/pull/1164#pullrequestreview-472375204) to it's own issue.
> Mergable afterwards.

Opened new issue for the mouse flickering #1170
